zanata-war/src/main/java/org/zanata/transaction/TransactionUtil.java

```diff
@@ -0,0 +1,162 @@
+            if (newTransactionRequired) {
+                throw e;
+            }
+        }
+
+        try {
+            if (newTransactionRequired) {
+                log.debug("beginning transaction");
+                userTransaction.begin();
+            }
+
+            R result = function.call();
+            if (newTransactionRequired) {
+                if (transaction.isMarkedRollback()) {
+                    log.debug("rolling back transaction");
+                    userTransaction.rollback();
+                } else {
+                    log.debug("committing transaction");
+                    userTransaction.commit();
+                }
+            }
+            return result;
+        } catch (Exception e) {
+            if (newTransactionRequired
+                    && userTransaction.getStatus() != Status.STATUS_NO_TRANSACTION) {
+                if (isRollbackRequired(e, true)) {
+                    log.debug("rolling back transaction");
+                    userTransaction.rollback();
+                } else {
+                    log.debug("committing transaction after ApplicationException(rollback=false):"
+                            + e.getMessage());
+                    userTransaction.commit();
+                }
+            }
+            throw e;
+        }
+    }
+
+    /**
+     * Same as {@link TransactionUtil#runInTransaction(Callable)} but for
+     * Runnables (functions that don't return anything)
+     *
+     * @param function
+     *            The function (in the form of a Runnable) to execute.
+     * @throws Exception
+     *             Exception (if any) thrown by the given function.
+     * @see TransactionUtil#runInTransaction(Callable)
+     */
+    public static void runInTransaction(Runnable function) throws Exception {
+        runInTransaction(() -> {
+            function.run();
+            return null;
+        });
+    }
+
+    protected static boolean
+            isNewTransactionRequired(boolean transactionActive) {
+        return !transactionActive;
+    }
+
+    public static boolean isRollbackRequired(Exception e, boolean isJavaBean) {
+        Class<? extends Exception> clazz = e.getClass();
+        return (isSystemException(e, isJavaBean, clazz))
+                ||
+                (isJavaBean && clazz.isAnnotationPresent(APPLICATION_EXCEPTION) && rollback(clazz
+                        .getAnnotation(APPLICATION_EXCEPTION)))
+                ||
+                (clazz.isAnnotationPresent(ApplicationException.class) && clazz
+                        .getAnnotation(ApplicationException.class).rollback());
+    }
+
+    private static boolean isSystemException(Exception e, boolean isJavaBean,
+            Class<? extends Exception> clazz) {
+        return isJavaBean &&
+                (e instanceof RuntimeException) &&
+                !clazz.isAnnotationPresent(APPLICATION_EXCEPTION) &&
+                !clazz.isAnnotationPresent(ApplicationException.class) &&
```

Comment:
Last time I checked, we don't have a case to use ApplicationException or APPLICATION_EXCEPTION, so this two conditions can be removed.